imi/src/main/java/org/motechproject/nms/imi/service/impl/CdrFileServiceImpl.java

```diff
@@ -346,6 +373,129 @@ private void processCsrEvent(CallSummaryRecordDto csrDto, boolean distributed) {
+    }
+
+
+    private String hostName() {
+        try {
+            return InetAddress.getLocalHost().getHostName();
+        } catch (UnknownHostException e) {
+            return "###";
+        }
+    }
+
+
+    private void reportIfAllChunksWereProcessed(final String file) {
+
+        @SuppressWarnings("unchecked")
+        SqlQueryExecution<List<ChunkAuditRecord>> queryExecution = new SqlQueryExecution<List<ChunkAuditRecord>>() {
+
+            @Override
+            public String getSqlQuery() {
+                String query = String.format("SELECT * FROM nms_imi_chunk_audit_record WHERE file = '%s' AND " +
+                        "node IS NULL LIMIT 1", file);
+                LOGGER.debug("SQL QUERY: {}", query);
+                return query;
+            }
+
+            @Override
+            public List<ChunkAuditRecord> execute(Query query) {
+
+                query.setClass(CallRetry.class);
+
+                ForwardQueryResult fqr = (ForwardQueryResult) query.execute();
+
+                return (List<ChunkAuditRecord>) fqr;
+            }
+        };
+
+        List<ChunkAuditRecord> records = chunkAuditRecordDataService.executeSQLQuery(queryExecution);
+
+        if (records.size() == 0) {
+            LOGGER.info("All chunk(s) of {} were processed.", file);
+        }
+    }
+
+
+    private void upsertChunkAuditRecord(String file, String chunk, int csrCount) {
+        ChunkAuditRecord record = chunkAuditRecordDataService.findByFileAndChunk(file, chunk);
+        if (record == null) {
+            chunkAuditRecordDataService.create(new ChunkAuditRecord(file, chunk, csrCount, 0, null));
+        } else {
+            record.setCsrProcessed(csrCount);
+            record.setNode(hostName());
+            chunkAuditRecordDataService.update(record);
+        }
+    }
+
+
+    private void updateChunkAuditRecord(String file, String chunk, int csrCount) {
+        ChunkAuditRecord record = chunkAuditRecordDataService.findByFileAndChunk(file, chunk);
+        if (record != null) {
+            record.setCsrProcessed(csrCount);
+            record.setNode(hostName());
+            chunkAuditRecordDataService.update(record);
+        }
+    }
+
+
+    @MotechListener(subjects = { NMS_IMI_PROCESS_CHUNK })
+    @Transactional
+    public void processChunk(MotechEvent event) {
+        Timer timer = new Timer("csr", "csrs");
+        String file = (String) event.getParameters().get("file");
+        String name = (String) event.getParameters().get("name");
+        String json = (String) event.getParameters().get("chunk");
+        ObjectMapper mapper = new ObjectMapper();
+        List<CallSummaryRecordDto> csrDtos;
+
+        try {
+            try {
+                csrDtos = mapper.readValue(json, new TypeReference<List<CallSummaryRecordDto>>() { });
+            } catch (IOException e) {
+                throw new ChunkingException(String.format("Exception unpacking packaging CSR chunk %s: %s", name,
```

Comment:
If nobody catches this specific exception type, then just let the bigger try catch-all and log? we can get rid of this inner try-catch?